cpp/src/tsne/barnes_hut.h

```diff
@@ -122,19 +117,45 @@ void Barnes_Hut(float *VAL, const int *COL, const int *ROW, const int NNZ,
 
   // Apply
   float *gains_bh = (float *)d_alloc->allocate(sizeof(float) * n * 2, stream);
-  thrust::device_ptr<float> begin_gains_bh =
-    thrust::device_pointer_cast(gains_bh);
-  thrust::fill(thrust::cuda::par.on(stream), begin_gains_bh,
-               begin_gains_bh + (n * 2), 1.0f);
+  thrust::fill(thrust::cuda::par.on(stream), gains_bh, gains_bh + (n * 2), 1.0f);
+  CUDA_CHECK(cudaPeekAtLastError());
 
   float *old_forces = (float *)d_alloc->allocate(sizeof(float) * n * 2, stream);
   CUDA_CHECK(cudaMemsetAsync(old_forces, 0, sizeof(float) * n * 2, stream));
 
-  float *YY =
-    (float *)d_alloc->allocate(sizeof(float) * (nnodes + 1) * 2, stream);
-  random_vector(YY, -0.0001f, 0.0001f, (nnodes + 1) * 2, stream, random_state);
+  float *YY = (float *)d_alloc->allocate(sizeof(float) * (nnodes + 1) * 2, stream);
   ASSERT(YY != NULL && rep_forces != NULL, "[ERROR] Possibly no more memory");
 
+
+  // Intialize embeddings
+  if (new_intialization == false) {
+
+    thrust::device_ptr<float> Y_ = thrust::device_pointer_cast(Y);
+    const float max = fabs(*thrust::max_element(thrust::cuda::par.on(stream), Y_, Y_ + n*2));
```

Comment:
These are being used enough in a few different algorithms that we should move both min & max to the prims. We should also see if our existing reduction primitive could be used to implement this.

Reply:
Yep thats a good point! I'll try moving these as a new primitive :)